### PR TITLE
feat(ext/net): implement Happy Eyeballs for `Deno.connect` and `Deno.connectTls`

### LINKED ISSUE #25661: Implement Happy Eyeballs for TCP connections
Right now we generally prefer IPv6 to IPv4 addresses when connecting to services. This unfortunately causes users to run into "connection failed" errors when their service does not actually listen on IPv6 (even though the DNS resolves to an IPv6 address).

[Happy Eyeballs](https://en.wikipedia.org/wiki/Happy_Eyeballs) fixes this. It establishes connections to multiple addresses in parallel and uses the first one that connects.

### PR DESCRIPTION
**Pull request by magurotuna**
Implements the Happy Eyeballs v2 ([RFC 8305]) for TCP and TLS connections, enabling faster connection establishment on dual-stack networks by racing IPv6 and IPv4 addresses with staggered parallel attempts.

[RFC 8305]: https://datatracker.ietf.org/doc/html/rfc8305

Closes #25661

## Changes

- Add `ext/net/happy_eyeballs.rs` with the core algorithm implementation
- Update `Deno.connect` and `Deno.connectTls` to use Happy Eyeballs when multiple addresses are resolved
- Add two new options to `Deno.ConnectOptions` and `Deno.ConnectTlsOptions`:
  - `autoSelectFamily`: Enable/disable Happy Eyeballs (default: `true`)
  - `autoSelectFamilyAttemptDelay`: Delay between connection attempts in ms (default: `250`)

## How it works

1. DNS resolution returns multiple addresses (potentially both IPv6 and IPv4)
2. Addresses are interleaved: IPv6 first, then IPv4, alternating (per [RFC 8305] Section 4)
3. First connection attempt starts immediately
4. New attempts start at configurable intervals (default: 250ms) while previous attempts continue in parallel
5. First successful connection wins; all other pending attempts are cancelled

This avoids the common problem where a broken IPv6 path causes long delays before falling back to IPv4.

## Default behavior

`autoSelectFamily` defaults to `true`, so Happy Eyeballs is on by default for `Deno.connect` and `Deno.connectTls`. This is a deliberate choice for parity with Node's `net.connect()`, whose `autoSelectFamily` default is also `true`, and it gives the better out-of-the-box experience on dual-stack networks where a broken IPv6 path would otherwise stall every connection. Callers who want the previous single-address behavior can opt out per-connection with `autoSelectFamily: false`.

Note that this changes the default connection-establishment semantics of a stable API: with multiple resolved addresses, which address wins the race (and therefore what `localAddr` reports) can differ from before. Single-address resolutions are unaffected in practice, since Happy Eyeballs is a no-op when only one address is resolved.

### CODE REVIEW

**Comment by magurotuna** [comment]
Node compat layer have already had its own Happy Eyeballs logic implemented in JS. So initially I thought it would make sense to implement it in Rust so that both `Deno.connect()` and Node's `net.connect()` can leave the Happy Eyeballs logic to the Rust side, which would allow us to remove it from JS. However, Node's implementation has several points where events are emitted such as `connectionAttempt` or `connectionAttemptFailed`, making it hard to perform the Happy Eyeballs algorithm entirely in Rust, as there's no mechanism to emit JS events from Rust.
https://nodejs.org/api/net.html#event-connectionattemptfailed

**Comment by Hajime-san** [comment]
> there's no mechanism to emit JS events from Rust.

Has this refactor made it available?

- https://github.com/denoland/deno/pull/30128

**Comment by bartlomieju** [comment]
> > there's no mechanism to emit JS events from Rust.
> 
> Has this refactor made it available?
> 
> * [refactor: object wrap Event #30128](https://github.com/denoland/deno/pull/30128)

Yes, but we benched it and at the moment it's 10x slower than emitting events from JS.

File: ext/net/happy_eyeballs.rs [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+          }
+          Err(e) => {
+            last_error = e;
+            // Continue waiting for other pending connections or start new ones
+          }
+        }
+      }
+
+      // Time to start next connection (staggered start per RFC 8305)
+      _ = sleep_until(next_attempt_at), if has_more_addrs => {
```
Comment (goffrie) [inline ext/net/happy_eyeballs.rs:107]: If `has_pending = false` & `has_more_addrs = true`, we should probably start the next attempt immediately instead of waiting; in particular if IPv6 is broken and fails immediately then we should quickly fall back to IPv4

Comment (magurotuna) [reply ext/net/happy_eyeballs.rs:107]: That's a good optimization! Implemented that in [e90a168](https://github.com/denoland/deno/pull/31726/commits/e90a168db737846292f674dbbe0de6fd37d19e7e)

**Review by kajukitli — COMMENTED** [review]
solid implementation of RFC 8305. a few observations:

**the good:**
- interleaving logic is correct (IPv6 first, alternating)
- staggered parallel attempts with configurable delay
- immediate fallback when no pending attempts (avoids waiting `attempt_delay` after fast failure)
- good test coverage including edge cases (cancellation, timeout fallback, parallel racing)
- clean API matching Node.js's `autoSelectFamily` options

**questions/suggestions:**

1. **cancel_handle not passed to TLS** — in `op_net_connect_tls`, you call `connect_happy_eyeballs(addrs, attempt_delay, None)` with `None` for cancel_handle, but `op_net_connect_tcp_inner` passes the actual cancel handle. is this intentional? TLS connections could be long-running and users might want to abort them.

2. **`get_refusing_addr()` in tests** — using port 0 as a refusing address is clever, but the comment says "Port 0 is not a valid destination port for TCP connects." this is true but the behavior is OS-dependent. on some systems it might error differently. might be worth a note or using a different approach like `192.0.2.1:1` (TEST-NET-1, guaranteed unroutable).

3. **single address path** — when `addrs.len() == 1`, you skip Happy Eyeballs entirely. makes sense for performance, but it means the `attempted_addresses` tracking won't happen. not sure if that matters for debugging/observability.

4. **error aggregation** — `last_error` only keeps the last failure. for debugging multi-address failures, it might be useful to aggregate errors or at least mention how many addresses were tried. low priority though.

5. **missing integration test** — the unit tests are great, but a real integration test with actual IPv4/IPv6 dual-stack would be nice. though I understand that's hard to set up in CI.

overall this looks good to ship. the core algorithm is correct and matches the RFC.

File: ext/net/happy_eyeballs.rs [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+      pending.push(start_connect(addr, cancel_handle.clone()));
+      next_attempt_at = Instant::now() + attempt_delay;
+      continue;
+    }
+
+    tokio::select! {
+      biased;
+
+      // A connection attempt completed
+      result = pending.next(), if has_pending => {
```
Comment (fibibot) [inline ext/net/happy_eyeballs.rs:90]: Subtle question on the `last_error` semantics: when all addresses fail, the user gets the error from the *last* attempt to fail (chronologically), not necessarily the one that's most actionable. If IPv6 fails fast with `EHOSTUNREACH` and IPv4 then fails slow with `ECONNREFUSED`, the user sees `ECONNREFUSED` only. That's probably fine and matches Node's behavior, but worth a one-line comment explaining the choice.

File: ext/net/01_net.js [hunk 3]
```diff
@@ -716,6 +716,11 @@ async function connect(args) {
             },
             undefined,
             cancelRid,
+            {
+              autoSelectFamily: args.autoSelectFamily ?? true,
```
Comment (fibibot) [inline ext/net/01_net.js:718]: `args.autoSelectFamily ?? true` and `args.autoSelectFamilyAttemptDelay ?? 250` use nullish-coalescing. That correctly preserves explicit `false` / `0`, but does flow `null` through to defaults — so `connect({autoSelectFamily: null})` enables it (rather than throwing or treating as falsy). Probably the intended semantics, just want to confirm WebIDL conformance isn't expected to reject null here.

File: ext/net/happy_eyeballs.rs [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+    assert_eq!(result.len(), 3);
+    assert!(result[0].is_ipv6());
+    assert!(result[1].is_ipv4());
+    assert!(result[2].is_ipv6());
+  }
+
+  #[test]
+  fn test_interleave_more_v4() {
+    let v6_1 = SocketAddr::V6(SocketAddrV6::new(Ipv6Addr::LOCALHOST, 80, 0, 0));
+    let v4_1 = SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::LOCALHOST, 80));
```
Comment (fibibot) [inline ext/net/happy_eyeballs.rs:258]: The `start_connect` cancellation-handling unwraps `?` twice (first for `or_cancel` mapping, second for the inner connect) — works but is dense. Worth one line splitting it for readability:

```rust
let cancelled = TcpStream::connect(addr).or_cancel(&cancel).await;
let stream = cancelled.map_err(|_| IoError::new(ErrorKind::Interrupted, "Connection cancelled"))??;
```

Not blocking — just a minor smell. (Not a real concern; mention only because the double `??` made me re-read.)

**Review by fibibot — COMMENTED** [review]
Substance is solid:

- **`interleave_addresses`** correctly partitions and zips IPv6/IPv4 to produce `[v6_1, v4_1, v6_2, v4_2, ...]` per RFC 8305 §4. Tests cover the empty / single-family / balanced / unbalanced cases.
- **`connect_happy_eyeballs`** races `FuturesUnordered` of `start_connect` futures with staggered starts, and crucially short-circuits the wait when there are no pending attempts (`if !has_pending && has_more_addrs` arm at lines 76-83) — so a fast TCP RST doesn't waste 250ms before trying the next address. `tokio::select! biased` ensures completed-attempt resolution is checked before the timer arm.
- **Cancellation**: every `start_connect` shares the same `Rc<CancelHandle>`, so an AbortSignal fans out to all in-flight attempts cleanly. The successful attempt's `TcpStream` is returned and the cancellation propagates via Drop on the dropped `FuturesUnordered`.
- **Default `attempt_delay = 250ms`** matches Node's `net.connect()` `autoSelectFamilyAttemptDelay` default and Chromium's Happy Eyeballs implementation.

CI is fully green (109 SUCCESS / 2 SKIPPED) on this HEAD.

Holding at COMMENT instead of APPROVE because **this changes default behavior of a stable Deno API** — `autoSelectFamily: true` is the new default for `Deno.connect` / `Deno.connectTls`, so even callers who don't opt in will see different connection-establishment semantics (which address wins the race, what `localAddr` reports, what timing-sensitive tests observe). That's on the human-escalation list per the review policy.

It's worth a maintainer's call on:

1. **Default value.** Should `autoSelectFamily` default to `true` (Node-parity) or `false` (preserve existing Deno behavior, opt-in only)?
2. **Behavior on single-address resolutions.** When DNS returns one address, Happy Eyeballs is a no-op except for the `attempted_addresses` bookkeeping. Worth a one-line confirmation that the perf overhead is negligible vs. a direct `TcpStream::connect`.
3. **`Deno.serve` / `Deno.listen` / Deno.connect over Unix sockets** are unaffected — the new options are documented as TCP-only via `transport?: "tcp"`. Worth confirming there's no path where a Unix-socket connect accidentally hits this code.

PR is also `dirty` against main from 2025-12-27 — almost 5 months stale. Rebase needed before merge regardless.

cc @bartlomieju.

**Comment by bartlomieju** [comment]
Rebased onto latest `main` and resolved the conflicts. Pushed as a merge commit (no force-push).

The non-trivial part of the merge: `main` recently landed a post-resolution deny check (`check_net_resolved`) on `Deno.connect` / `Deno.connectTls`, which prevents bypassing IP-literal deny rules via numeric hostname aliases (e.g. `2130706433` -> `127.0.0.1`). That check collided with the Happy Eyeballs path:

- For `Deno.connect` (`ext/net/ops.rs`) git produced a *textually* clean auto-merge that was actually broken: it called `check_net_resolved(&addr.ip(), addr.port(), ...)` where `addr` is the `IpAddr { hostname, port }` op argument, which has no `.ip()` method (would not compile).
- For `Deno.connectTls` (`ext/net/ops_tls.rs`) the Happy Eyeballs version had dropped the deny check entirely.

Resolved both by running the deny check over **all** resolved candidate addresses before connecting, since Happy Eyeballs may connect to any of them (failing closed if any resolved IP is denied). Also re-added the `use serde::Deserialize;` import that the merge had dropped (needed by `TcpConnectOptions`).

Addressing the earlier review threads:

- **@goffrie** (immediate fallback when no attempt is pending) — already implemented in e90a168; the `!has_pending && has_more_addrs` arm starts the next attempt immediately instead of waiting `attempt_delay`.
- **@fibibot** (`last_error` semantics) — added a comment noting we surface the chronologically-last failure, matching Node's `net.connect()`.
- **@kajukitli** (cancel handle not passed to TLS) — the `op_net_connect_tls` op has no abort-resource parameter at all, so there is no cancel handle to pass; `None` matches the pre-Happy-Eyeballs behavior (TLS connect was never cancellable). Wiring cancellation into TLS would be a separate change.
- **@kajukitli** (port 0 as a refusing address in tests) — kept; it gives a deterministic *fast* connection refusal, which is what the fast-fallback timing assertions rely on. A non-routable TEST-NET address would hang rather than refuse and break those tests.

Local `cargo test -p deno_net happy_eyeballs` is green (17/17). Leaving the rest to CI.

File: ext/net/happy_eyeballs.rs (hunk 2 above)
Comment (bartlomieju) [reply ext/net/happy_eyeballs.rs:90]: Addressed — there is now a comment where `last_error` is initialized in `connect_happy_eyeballs` explaining the choice: when every attempt fails we surface the error from the last attempt to fail (chronologically), which matches Node `net.connect()` behavior of reporting the final failed attempt.

File: ext/net/01_net.js (hunk 3 above)
Comment (bartlomieju) [reply ext/net/01_net.js:718]: Confirmed intentional. `Deno.connect` / `Deno.connectTls` are Deno-namespace APIs, not WebIDL-defined Web platform interfaces, so there is no conformance requirement to reject `null` here. The `?? true` / `?? 250` semantics deliberately treat both `null` and `undefined` as "use the default" while preserving an explicit `false` / `0`, which matches Node `autoSelectFamily` / `autoSelectFamilyAttemptDelay` handling.

File: ext/net/happy_eyeballs.rs (hunk 4 above)
Comment (bartlomieju) [reply ext/net/happy_eyeballs.rs:258]: Done in 065882a018 — the nested `?` is now split across separate lines with a comment noting the outer result is cancellation and the inner one is the connect itself.

File: ext/net/ops.rs [hunk 5]
```diff
@@ -512,19 +544,30 @@ pub async fn op_net_connect_tcp_inner(
...
+  // aliases (e.g. 2130706433 → 127.0.0.1). All candidate addresses are
+  // checked because Happy Eyeballs may connect to any of them.
   {
-    state
-      .borrow_mut()
-      .borrow_mut::<PermissionsContainer>()
-      .check_net_resolved(&addr.ip(), addr.port(), "Deno.connect()")?;
+    let mut state_ = state.borrow_mut();
+    let permissions = state_.borrow_mut::<PermissionsContainer>();
+    for addr in &addrs {
```
Comment (bartlomieju) [inline ext/net/ops.rs]: This now denies the connect if *any* resolved address is denied, even though the non-HE branch below only connects to `addrs[0]`. Previously the check was on just the single used address, so a hostname resolving to `[allowed, denied]` could connect to the allowed one; now it fails closed. Correct for the HE race, but a behavior change for `autoSelectFamily: false` / single-attempt. Same pattern in `ext/net/ops_tls.rs`. Suggest gating the all-candidates check on the HE branch and checking only the attempted address otherwise. Follow-up, not blocking.

**Review by bartlomieju — COMMENTED** [review]
Flagging one behavior change to address as a follow-up (not blocking this landing).

The post-resolution deny check now runs `check_net_resolved` over **all** resolved candidate addresses and fails closed if any one is denied. For the Happy Eyeballs path that's exactly right (the connect may land on any candidate). But it also applies when HE doesn't actually race them:

- when `autoSelectFamily: false`, and
- in general the op only connects to `addrs[0]` in the non-HE branch,

yet the check still rejects if *any other* resolved address is denied. So a hostname resolving to e.g. `[allowed_v4, denied_v6]` under `--deny-net=<v6>` now fails the whole connect, where previously (check on just the first/used address) it succeeded. That's a previously-working `Deno.connect` / `Deno.connectTls` now erroring.

"Deny wins across all candidates" is a defensible stance, but it's a real, potentially surprising regression for the non-racing paths, and it affects both `op_net_connect_tcp_inner` (`ext/net/ops.rs`) and `op_net_connect_tls` (`ext/net/ops_tls.rs`). Likely fix: scope the all-candidates check to the branch that actually races (HE enabled + `addrs.len() > 1`), and check only the address(es) we'll really attempt otherwise.

Good to land as-is; let's track this and fix it in a follow-up.

File: ext/net/ops.rs (hunk 5 above)
Comment (bartlomieju) [reply ext/net/ops.rs]: Addressed in aed16c4854. The all-candidates deny check is now gated on the branch that actually races (`use_happy_eyeballs = auto_select_family && addrs.len() > 1`); the non-racing path checks only `addrs[0]`, the single address it will connect to. So a hostname resolving to `[allowed, denied]` under `--deny-net=<denied>` connects to the allowed address again with `autoSelectFamily: false` or a single resolution, while the HE race still fails closed if any candidate it may attempt is denied. Same fix applied to `op_net_connect_tls`. Numeric-alias deny spec tests (`deny_net_numeric_hostname_subnet`, `tcp_numeric_deny`) still pass.
